### DISCUSSION TITLE
Guidance on setting up a development environment?

### DISCUSSION THREAD

**Discussion by zmbc** [moj-analytical-services/splink#2019] [Q&A] [answered]
I'm currently trying to set up an environment for working on Splink for the first time, and am struggling a bit.

The [contributing code section of CONTRIBUTING.md](https://github.com/moj-analytical-services/splink/blob/master/CONTRIBUTING.md#contributing-code) doesn't really have a quickstart. I followed the link to [Creating a Virtual Environment with Splink](https://github.com/moj-analytical-services/splink/blob/master/docs/dev_guides/changing_splink/building_env_locally.md), but it focuses more on how to _update_ the environment (though I did glean that I need to run `poetry install`).

It then links to [the create_venv script](https://github.com/moj-analytical-services/splink/blob/master/scripts/create_venv.sh), which installs poetry itself and Splink's dependencies in the same virtual environment. This is [not recommended](https://github.com/python-poetry/poetry/issues/7878#issuecomment-1535496388) and I was not able to get this to work on my Windows Subsystem for Linux setup. The default behavior of `poetry install` is to create a _new_ venv, so I am doing that instead.

I am also working inside a conda environment, and had to work around [Poetry's default behavior being wrong in a conda environment](https://github.com/python-poetry/poetry/issues/4055) with [this plugin](https://github.com/renan-r-santos/poetry-conda).

At this point, I am able to run the tests, and some of them pass 🎉 However, it seems the tests also depend on having Postgres set up in a particular way, even though I don't see where that is done [in the CI](https://github.com/moj-analytical-services/splink/blob/8b865c3cf0e4678d35ac7b0c46dd85ddcc8ced6d/.github/workflows/pytest_run_tests_with_cache.yml).

This looks to be the relevant section on how to set up Postgres, but it doesn't seem to be documented anywhere:

https://github.com/moj-analytical-services/splink/blob/8b865c3cf0e4678d35ac7b0c46dd85ddcc8ced6d/tests/backend_utils/postgres_conf.py#L13-L19

Lastly, I do not understand how Spark is getting set up for the tests. Unlike Postgres, it [is not included in the base Ubuntu image the CI uses](https://github.com/actions/runner-images/blob/ubuntu20/20240225.1/images/ubuntu/Ubuntu2004-Readme.md). When I run the tests locally, I just get lots of `py4j.protocol.Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.` errors, which is not surprising since I haven't installed Spark.

Opening this as a discussion to see if I am missing anything obvious on any of these points -- but if not, perhaps my first contribution to Splink can be documenting the development setup?

**Comment by ADBond** [comment]
> It then links to [the create_venv script](https://github.com/moj-analytical-services/splink/blob/master/scripts/create_venv.sh), which installs poetry itself and Splink's dependencies in the same virtual environment. This is [not recommended](https://github.com/python-poetry/poetry/issues/7878#issuecomment-1535496388) and I was not able to get this to work on my Windows Subsystem for Linux setup. The default behavior of `poetry install` is to create a _new_ venv, so I am doing that instead.

You are right - that sounds like an error in the script. Working in a fresh env (not where poetry is installed) is the preferred thing.

> I am also working inside a conda environment, and had to work around [Poetry's default behavior being wrong in a conda environment](https://github.com/python-poetry/poetry/issues/4055) with [this plugin](https://github.com/renan-r-santos/poetry-conda).

This is useful to know, thanks! We don't tend to use `conda` in the team, so useful to know difficulties with it.

> At this point, I am able to run the tests, and some of them pass 🎉 However, it seems the tests also depend on having Postgres set up in a particular way, even though I don't see where that is done [in the CI](https://github.com/moj-analytical-services/splink/blob/8b865c3cf0e4678d35ac7b0c46dd85ddcc8ced6d/.github/workflows/pytest_run_tests_with_cache.yml).

The non-postgres tests _should_ work okay if you don't have postgres installed/setup, but there may well be something wrong with our setup on that front. If you could share the errors/failures you get with this that would be a great help. The postgres tests themselves are run in [a separate workflow](https://github.com/moj-analytical-services/splink/blob/master/.github/workflows/pytest_postgres.yml). For running locally there is [this folder](https://github.com/moj-analytical-services/splink/tree/master/scripts/postgres) which has a docker-compose file to set up a container running postgres (configured to work with the tests)

> This looks to be the relevant section on how to set up Postgres, but it doesn't seem to be documented anywhere:

There is [this page on postgres](https://moj-analytical-services.github.io/splink/topic_guides/splink_fundamentals/backends/postgres.html) which has some information, but probably not easy to discover, and may well be lacking in relevant details.

> Lastly, I do not understand how Spark is getting set up for the tests. Unlike Postgres, it [is not included in the base Ubuntu image the CI uses](https://github.com/actions/runner-images/blob/ubuntu20/20240225.1/images/ubuntu/Ubuntu2004-Readme.md). When I run the tests locally, I just get lots of `py4j.protocol.Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.` errors, which is not surprising since I haven't installed Spark.

Spark is installed automatically with `pyspark`, which is how it works in the CI. However this is something that appears to not always 'just work' for people - I'm no expert on this, but if that causes trouble you can try to [manually install Spark](https://spark.apache.org/downloads.html)

> Opening this as a discussion to see if I am missing anything obvious on any of these points -- but if not, perhaps my first contribution to Splink can be documenting the development setup?

We'd certainly welcome any contribution on documentation for elements of the dev setup that you are think are unclear/lacking/wrong - as someone coming in with fresh eyes you're probably in a good position to know what would be most useful to add.

**Reply by zmbc** [reply]
> For running locally there is [this folder](https://github.com/moj-analytical-services/splink/tree/master/scripts/postgres) which has a docker-compose file to set up a container running postgres

My assumption is that it should always be possible to run the tests locally -- otherwise I won't catch bugs until I push my changes and run on the CI (which isn't even easy on a fork, if I recall correctly). And personally I've never worked on a project where the normal way to run the tests locally was inside of containers -- that adds a lot of overhead (and again, slows down the iteration cycle). But if that is what Splink intends, I am happy to try to get that working and then document it better.

> Spark is installed automatically with `pyspark`

I may be misunderstanding what you mean, but this is not my experience. The PyPI description for PySpark says:

> The Python packaging for Spark is not intended to replace all of the other use cases. **This Python packaged version of Spark is suitable for interacting with an existing cluster** (be it Spark standalone, YARN, or Mesos) - but does not contain the tools required to set up your own standalone Spark cluster. **You can download the full version of Spark from** the [Apache Spark downloads page](https://spark.apache.org/downloads.html).

Additionally, Spark depends on Java, which can't be installed through PyPI IIUC.

**Reply by zmbc** [reply]
Hm, on a bit more digging, it looks like pyspark automatically installs the Spark JARs. But you do have to have Java installed, or it won't work. And the Splink Spark tests may require a bit more setup that I haven't done.

**Reply by RobinL** [reply]
@zmbc here's a full step by step guide to getting everything (except postgres tests) set up in Windows/WSL:

https://gist.github.com/RobinL/7e52049e695e79a0b388be3169f7e4d7

Getting posgres tests passing locally requires having postgres running in Docker as Andy linked to [here](https://moj-analytical-services.github.io/splink/topic_guides/splink_fundamentals/backends/postgres.html#postgres-via-docker)

Unless I'm working on postgres-specific code, I personally don't usually bother and I would just let the postgres tests run in CI

**Reply by zmbc** [reply]
@RobinL Thanks so much!

Would you be open to a PR that does the following?
- Incorporates this into the documentation, making it a bit less Windows/WSL-specific
- Includes an option using conda, which is necessary if you don't have root access or don't want to install Python/Java globally

**Reply by zmbc** [reply]
Lastly, I also think it would be nice (though apparently it hasn't been a big issue for you all) to run the Postgres tests without using Docker containers (which require root access). I believe this could be achieved pretty easily with conda.

**Reply by zmbc** [reply]
@RobinL just want to check -- does this sound like a PR you would consider? I've done a little bit of work on it and have all tests (including Postgres) running now in a conda environment, but don't want to go to the trouble to clean it up/document it unless the overall idea sounds right to you. Thanks!

**Reply by RobinL** [reply]
Hi @zmbc .  Sorry for the delay.  Yes, absolutely would be open to a PR, I think it'd be very useful if you have the time.

If you'd like to go ahead, it might be worth just jotting down a very quick summary in bullet points of the planned content (i'm thinking like 10 bullet points).  Just so we can give it a quick skim and a thumbs up before you go ahead and put more time into it.  Thank you!

**Reply by zmbc** [reply]
Here's a rough plan: replace [the current "Building a Virtual Environment" page](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/building_env_locally.html) with two pages: **Development Quickstart** and **Managing Dependencies**.

**Development Quickstart**
- Option A: Python venv
  - Manual steps to install Python (platform specific)
  - Manual steps to install Java (platform specific)
  - Manual steps to install and set up Postgres (platform specific) (optional)
  - Manual step to install Poetry
  - `poetry install`
  - `poetry shell`
- Option B: automatic conda setup
  - Script that can run on any Unix-like, will install miniforge, install Python, Java, Postgres etc into an environment, set up the splink testing database in Postgres, then `poetry install`
  - `mamba activate splink`
  - `poetry shell`
- Option C: manual conda setup
  - Manual steps replicating the script (individual commands to run in terminal)
  - `mamba activate splink`
  - `poetry shell`

Note that this is the setup I've found necessary just to run the tests (which seems like a pretty good check that I've set things up correctly). However, there might be more that is needed and I don't know about. As one small example, I noticed that the spellchecker is its own thing with its own dependencies and currently doesn't seem to be cross-platform.

**Managing Dependencies**

I'd include here the content currently on the Virtual Environment page that is about adding/locking/etc as this isn't relevant to initial setup.

**Reply by RobinL** [reply]
Brilliant - sounds great - thank you!

**Reply by zmbc** [reply]
@RobinL I realized there is one more thing I'm unsure about -- should I be contributing to the `master` branch or `splink4_dev` for a docs change like this? I didn't immediately find any guidance about what goes where.

**Reply by RobinL** [reply]
@zmbc good question. Let's put it on master to make it available as quickly as possible, and then I'll also merge it up into Splink 4 too.

**Reply by zmbc** [reply, marked as answer]
Pull request: https://github.com/moj-analytical-services/splink/pull/2083
